src/BinSkim.Driver/RuleTelemetryLogger.cs

```diff
@@ -0,0 +1,179 @@
+        }
+
+        public void AnalysisStarted()
+        {
+        }
+
+        public void AnalysisStopped(RuntimeConditions runtimeConditions)
+        {
+            EventTelemetry stopEvent = new EventTelemetry(AnalysisStoppedEventName);
+            stopEvent.Properties[RuntimeConditionsPropertyName] = runtimeConditions.ToString();
+            this.telemetryClient.TrackEvent(stopEvent);
+
+            foreach (KeyValuePair<string, ResultKindCounts> kv in this.metricsMap)
+            {
+                EventTelemetry eventTelemetry = new EventTelemetry(RuleSummaryEventName);
+                eventTelemetry.Properties[RuleIdPropertyName] = kv.Key;
+
+                // To reduce telemetry volume, record only non-zero values
+                ResultKindCounts counts = kv.Value;
+                IDictionary<string, double> metrics = eventTelemetry.Metrics;
+                AddMetricIfNonZero(metrics, counts.NotApplicableCount, NotApplicableCountMetricName);
+                AddMetricIfNonZero(metrics, counts.PassCount, PassCountMetricName);
+                AddMetricIfNonZero(metrics, counts.FailCount, FailCountMetricName);
+                AddMetricIfNonZero(metrics, counts.ReviewCount, ReviewCountMetricName);
+                AddMetricIfNonZero(metrics, counts.OpenCount, OpenCountMetricName);
+                AddMetricIfNonZero(metrics, counts.InformationalCount, InformationalCountMetricName);
+
+                this.telemetryClient.TrackEvent(eventTelemetry);
+            }
+
+            this.metricsMap.Clear();
+        }
+
+        public void AnalyzingTarget(IAnalysisContext context)
+        {
+        }
+
+        public void Log(ReportingDescriptor rule, Result result)
+        {
+            if (!this.metricsMap.TryGetValue(rule.Id, out ResultKindCounts? counts))
+            {
+                this.metricsMap[rule.Id] = counts = new ResultKindCounts();
+            }
+
+            switch (result.Kind)
+            {
+                case ResultKind.NotApplicable:
+                    counts.NotApplicableCount++;
+                    break;
+
+                case ResultKind.Pass:
+                    counts.PassCount++;
+                    break;
+
+                case ResultKind.Fail:
+                    counts.FailCount++;
+                    break;
+
+                case ResultKind.Review:
+                    counts.ReviewCount++;
+                    break;
+
+                case ResultKind.Open:
+                    counts.OpenCount++;
+                    break;
+
+                case ResultKind.Informational:
+                    counts.InformationalCount++;
+                    break;
+
+                default:
+                    break;
+            }
+        }
+
+        public void LogConfigurationNotification(Sarif.Notification notification)
+        {
+        }
+
+        public void LogToolNotification(Sarif.Notification notification)
```

Comment:
>LogToolNotification [](http://example.com/codeflow?start=20&length=19)

We don't need to block this first PR, but we will want to add some events on notifications we receive, which will include unhandled exceptions in the engine and individual rules. We will have a special interest in enabling AI telemetry to tell us if any specific rule crashed and was disabled. Doing this is blocked on a change we just made to the SARIF SDK (to add the associated rule descriptor to this method), so let's defer. Maybe I can drop this in and run past you for review, so I can continue to learn about what you've done. I think it'd provide a nice verification on behavior for your new rule (and others).

Reply:
Happy to help out. Yes, if "exception in rule x" goes through here, then it would be easy to log it. Although, I think the CompilerDataLogger already does something like that.

Reply:
Right, of course! The compiler data logger has a very undesirable behavior: on analysis completion it actually reparses the output file to collect metrics for a summary.  Not good. All of that logic should be replaced with precisely the pattern you've started here: a logger should collect relevant metrics during analysis and then blat them out in the telemetry channel on analysis completion.

We definitely need to get an updated to the SDK that I just checked in to finish cleaning this up. If you're still investing in this PR, feel free to merge from main and get that breaking change (adds a reporting descriptor for the associated rule to the Log(Notification) method. I also think it'd be just fine to polish this up very slightly, merge, and then take the next incremental change.